src/engine/runtime.js

```diff
@@ -2414,30 +2424,97 @@ class Runtime extends EventEmitter {
+     * @property {string} color - the color for the monitor for the given block id
+     * @property {string} label - the label for the monitor for the given block id
+     */
+    getMonitorLabelForBlock (id) {
+        // Having dynamic extension blocks makes it so that a single extension category can have
+        // multiple toolbox blocks with the same opcode. This previously wasn't possible.
+        // This means that we can no longer use just the opcode to look up the block and
+        // should instead look up the block using its unique id.
+        const block = this.flyoutBlocks.getBlock(id);
+        // If the block doesn't actually exist in the flyout, we should not
+        // be creating a label for it.
+        if (!block) return;
+
+        // Look up category info from the block
+        const extendedOpcode = block.opcode;
         const [category, opcode] = StringUtil.splitFirst(extendedOpcode, '_');
         if (!(category && opcode)) return;
 
         const categoryInfo = this._blockInfo.find(ci => ci.id === category);
         if (!categoryInfo) return;
 
-        const block = categoryInfo.blocks.find(b => b.info.opcode === opcode);
-        if (!block) return;
+        const isDynamicExtensionBlock =
+            block.mutation &&
+            block.mutation.blockInfo &&
+            block.mutation.blockInfo.isDynamic;
+        let extensionBlockInfo;
+        if (isDynamicExtensionBlock) {
+            extensionBlockInfo = block.mutation.blockInfo;
+        } else {
+            const extensionBlockDesc = categoryInfo.blocks.find(b => b.info.opcode === opcode);
+            extensionBlockInfo = (extensionBlockDesc && extensionBlockDesc.info) || null;
+        }
+        if (!extensionBlockInfo) return;
+
+        // TODO get rid of this when we update the core extension name
+        const temporaryCategory = category === 'data2' ? 'data' : category;
+        if (this.isCoreExtension(temporaryCategory)) {
+            return {
+                color: extensionBlockInfo.color1 || categoryInfo.color1,
+                label: extensionBlockInfo.text
+            };
+        }
 
         // TODO: we may want to format the label in a locale-specific way.
         return {
-            category: 'extension', // This assumes that all extensions have the same monitor color.
-            label: `${categoryInfo.name}: ${block.info.text}`
+            color: categoryInfo.color1, // TODO This assumes that all extensions have the same monitor color.
+            label: `${categoryInfo.name}: ${extensionBlockInfo.text}`
         };
     }
 
+    /**
+     * Keep track of the fact that a monitor has been requested
+     * for the given blockId, even though the block with the given id
+     * does not exist in the flyout yet.
+     * @param {string} blockId The id of the block with a pending monitor
+     */
+    addPendingMonitor (blockId) {
+        this._pendingMonitors[blockId] = true;
```

Comment:
If you're feeling fancy you could use [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) instead